plugins/woocommerce/includes/rest-api/Controllers/Version3/class-wc-rest-products-controller.php

```diff
@@ -1462,16 +1480,45 @@ public function get_collection_params() {
 	 */
 	protected function get_product_data( $product, $context = 'view' ) {
 		$data = parent::get_product_data( ...func_get_args() );
-		// Add stock_status if needed.
+
 		if ( isset( $this->request ) ) {
 			$fields = $this->get_fields_for_response( $this->request );
-			if ( in_array( 'stock_status', $fields ) ) {
+
+			// Add stock_status if needed.
+			if ( in_array( 'stock_status', $fields, true ) ) {
 				$data['stock_status'] = $product->get_stock_status( $context );
 			}
-			if ( in_array( 'has_options', $fields ) ) {
+
+			// Add has_options if needed.
+			if ( in_array( 'has_options', $fields, true ) ) {
 				$data['has_options'] = $product->has_options( $context );
 			}
+
+			$post_type_obj = get_post_type_object( $this->post_type );
+			if ( is_post_type_viewable( $post_type_obj ) && $post_type_obj->public ) {
+				$permalink_template_requested = in_array( 'permalink_template', $fields, true );
+				$generated_slug_requested     = in_array( 'generated_slug', $fields, true );
+
+				if ( $permalink_template_requested || $generated_slug_requested ) {
+					if ( ! function_exists( 'get_sample_permalink' ) ) {
+						require_once ABSPATH . 'wp-admin/includes/post.php';
+					}
+
+					$sample_permalink = get_sample_permalink( $product->get_id(), $product->get_name(), '' );
+
+					// Add permalink_template if needed.
+					if ( $permalink_template_requested ) {
```

Comment:
Nitpicky, but we could probably move this `if`s outside of the above statement to avoid deeply nesting.

Reply:
Where do you think I should move the `if`? I'm not seeing it!

Reply:
https://github.com/ROBERLANDIO/-octaedro-approved-this-pull-request.-Nice-job-joshuatf-This-is-testing-well-and-the-code-

Reply:
Looked at this again and still don't see a clean way to reduce the nesting while keeping the readability. Going to merge this PR as is, but I'm still curious in knowing how you'd structure the code!